File: backend/app/api/v1/projects.py
# Copyright 2026 SciLoop contributors
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
"""项目容器 API（WP09；``contracts.api_contract.key_endpoints.projects`` / 附录 B.5）。

======== ===================  ==========  ================================================
方法      路径                  鉴权        说明
======== ===================  ==========  ================================================
GET      ``/projects``          公开        项目列表（``is_demo`` 排序优先，分页）
POST     ``/projects``          Owner       创建项目
GET      ``/projects/{id}``    公开        项目详情（含迭代历史：pipeline_runs 逐轮）
======== ===================  ==========  ================================================

列表响应形状遵循 ``contracts.api_contract.pagination``：``{items,total,page,page_size}``；
错误形状遵循 ``contracts.api_contract.error_shape``：``{code,message,detail}``。

本模块**由 ``app.main.ROUTER_REGISTRY`` 第 9 项登记，模块出现即自动挂载**，
因此本文件不得改动 ``main.py``。
"""

from __future__ import annotations

import logging
from typing import Annotated, Any

from fastapi import APIRouter, Body, Depends, HTTPException, Query, status
from sqlalchemy import func, select

from app.core.security import require_owner
from app.db.models import Idea, PipelineRun, Project, Taskbook
from app.db.session import AsyncSessionLocal
from app.services.pipeline.stages import STAGE_ORDER

logger = logging.getLogger("sciloop.api.projects")

router = APIRouter(prefix="/projects", tags=["projects"])

OwnerDep = Annotated[None, Depends(require_owner)]

#: 项目状态取值（``contracts.enums.project_status``）
PROJECT_STATUSES: tuple[str, ...] = (
    "DRAFT",
    "TASKBOOK_LOCKED",
    "RUNNING",
    "RISK_EVALUATING",
    "WAIT_HUMAN",
    "CIRCUIT_BREAK",
    "REVIEWING",
    "DONE",
    "ABORTED",
)

MODES: tuple[str, ...] = ("manual", "auto")

MAX_PAGE_SIZE = 100


def _session_factory() -> Any:
    if AsyncSessionLocal is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail={"code": "db_unavailable", "message": "数据库会话不可用", "detail": None},
        )
    return AsyncSessionLocal


def _error(http_status: int, code: str, message: str, detail: Any = None) -> HTTPException:
    return HTTPException(
        status_code=http_status,
        detail={"code": code, "message": message, "detail": detail},
    )


def _iso(value: Any) -> str | None:
    return value.isoformat() if value is not None else None


def _project_dict(project: Project, *, include_settings: bool = True) -> dict[str, Any]:
    """项目基础字段（``is_demo`` / ``taskbook_id`` 如实返回，缺值置 null 不编造）。"""
    payload: dict[str, Any] = {
        "id": int(project.id),
        "name": project.name,
        "status": project.status,
        "mode": project.mode,
        "current_iteration": int(project.current_iteration or 0),
        "is_demo": bool(project.is_demo),
        "idea_id": int(project.idea_id) if project.idea_id is not None else None,
        "taskbook_id": int(project.taskbook_id) if project.taskbook_id is not None else None,
        "created_at": _iso(project.created_at),
        "updated_at": _iso(project.updated_at),
    }
    if include_settings:
        payload["settings"] = project.settings
    return payload


def _run_dict(run: PipelineRun) -> dict[str, Any]:
    return {
        "pipeline_run_id": int(run.id),
        "project_id": int(run.project_id),
        "iteration": int(run.iteration),
        "mode": run.mode,
        "status": run.status,
        "stop_reason": run.stop_reason,
        "total_cost_usd": float(run.total_cost_usd or 0),
        "started_at": _iso(run.started_at),
        "finished_at": _iso(run.finished_at),
        "created_at": _iso(run.created_at),
    }


# --------------------------------------------------------------------------- #
# 读操作（公开）
# --------------------------------------------------------------------------- #
@router.get("", summary="项目列表（is_demo 优先，公开只读）")
async def list_projects(
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=MAX_PAGE_SIZE)] = 20,
    status_filter: Annotated[str | None, Query(alias="status", max_length=24)] = None,
    is_demo: Annotated[bool | None, Query()] = None,
) -> dict[str, Any]:
    """项目列表：``is_demo DESC NULLS LAST, id ASC``（附录 B.5「``is_demo`` 排序优先」）。"""
    if status_filter is not None and status_filter not in PROJECT_STATUSES:
        raise _error(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "validation_error",
            f"status 非法：{status_filter}",
            {"allowed": list(PROJECT_STATUSES)},
        )

    filters = []
    if status_filter is not None:
        filters.append(Project.status == status_filter)
    if is_demo is not None:
        filters.append(Project.is_demo.is_(is_demo))

    async with _session_factory()() as session:
        total = (
            await session.execute(select(func.count()).select_from(Project).where(*filters))
        ).scalar_one()
        rows = (
            (
                await session.execute(
                    select(Project)
                    .where(*filters)
                    .order_by(Project.is_demo.desc().nulls_last(), Project.id.asc())
                    .offset((page - 1) * page_size)
                    .limit(page_size)
                )
            )
            .scalars()
            .all()
        )
        items = [_project_dict(row, include_settings=False) for row in rows]

    return {
        "items": items,
        "total": int(total or 0),
        "page": page,
        "page_size": page_size,
        "sort_by": "is_demo DESC NULLS LAST, id ASC",
    }


@router.get("/{project_id}", summary="项目详情（含迭代历史，公开只读）")
async def get_project(project_id: int) -> dict[str, Any]:
    """项目详情 + ``pipeline_runs`` 迭代历史（按 iteration 升序，如实返回 ``stop_reason``）。"""
    async with _session_factory()() as session:
        project = (
            await session.execute(select(Project).where(Project.id == project_id))
        ).scalar_one_or_none()
        if project is None:
            raise _error(
                status.HTTP_404_NOT_FOUND, "project_not_found", f"project {project_id} 不存在"
            )
        runs = (
            (
                await session.execute(
                    select(PipelineRun)
                    .where(PipelineRun.project_id == project_id)
                    .order_by(PipelineRun.iteration.asc(), PipelineRun.id.asc())
                )
            )
            .scalars()
            .all()
        )
        payload = _project_dict(project)
        payload["stage_order"] = list(STAGE_ORDER)
        payload["iteration_history"] = [_run_dict(run) for run in runs]
        payload["run_count"] = len(runs)
        latest = runs[-1] if runs else None
        payload["latest_run"] = _run_dict(latest) if latest is not None else None

    return payload


# --------------------------------------------------------------------------- #
# 写操作（Owner）
# --------------------------------------------------------------------------- #
@router.post("", status_code=status.HTTP_201_CREATED, summary="创建项目（Owner）")
async def create_project(
    _owner: OwnerDep,
    payload: Annotated[dict[str, Any], Body(...)],
) -> dict[str, Any]:
    """创建项目（``status=DRAFT``，``current_iteration=0``）。

    允许字段：``name``（必填）、``mode``（``manual``/``auto``，默认 ``manual``）、
    ``idea_id`` / ``taskbook_id``（可选外键）、``settings``（JSONB）、``is_demo``。
    """
    name = str(payload.get("name") or "").strip()
    if not name:
        raise _error(status.HTTP_422_UNPROCESSABLE_ENTITY, "validation_error", "name 必填", payload)
    if len(name) > 200:
        raise _error(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "validation_error",
            "name 长度不得超过 200",
            {"length": len(name)},
        )

    mode = str(payload.get("mode") or "manual")
    if mode not in MODES:
        raise _error(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "validation_error",
            f"mode 必须为 {list(MODES)} 之一",
            {"mode": mode},
        )

    settings = payload.get("settings")
    if settings is not None and not isinstance(settings, dict):
        raise _error(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "validation_error",
            "settings 必须为 JSON 对象",
            {"settings": settings},
        )

    idea_id = payload.get("idea_id")
    taskbook_id = payload.get("taskbook_id")
    for label, value in (("idea_id", idea_id), ("taskbook_id", taskbook_id)):
        if value is not None and not isinstance(value, int):
            raise _error(
                status.HTTP_422_UNPROCESSABLE_ENTITY,
                "validation_error",
                f"{label} 必须为整数或 null",
                {label: value},
            )

    async with _session_factory()() as session:
        # 外键存在性校验：宁可 422 也不让 IntegrityError 变成 500
        if idea_id is not None:
            exists = (
                await session.execute(select(Idea.id).where(Idea.id == int(idea_id)))
            ).scalar_one_or_none()
            if exists is None:
                raise _error(
                    status.HTTP_422_UNPROCESSABLE_ENTITY,
                    "validation_error",
                    f"idea_id={idea_id} 不存在（``ideas`` 无此记录）",
                )
        if taskbook_id is not None:
            exists = (
                await session.execute(select(Taskbook.id).where(Taskbook.id == int(taskbook_id)))
            ).scalar_one_or_none()
            if exists is None:
                raise _error(
                    status.HTTP_422_UNPROCESSABLE_ENTITY,
                    "validation_error",
                    f"taskbook_id={taskbook_id} 不存在（``taskbooks`` 无此记录）",
                )

        project = Project(
            name=name,
            mode=mode,
            status="DRAFT",
            current_iteration=0,
            settings=settings,
            is_demo=bool(payload.get("is_demo", False)),
            idea_id=int(idea_id) if idea_id is not None else None,
            taskbook_id=int(taskbook_id) if taskbook_id is not None else None,
        )
        session.add(project)
        await session.flush()
        result = _project_dict(project)
        result["stage_order"] = list(STAGE_ORDER)
        result["note"] = (
            "项目已创建（status=DRAFT）；锁定任务书后调用 POST /pipelines/{pid}/run 启动"
        )
        await session.commit()

    logger.info("project_created project_id=%s name=%s mode=%s", result["id"], name, mode)
    return result


@router.patch("/{project_id}", summary="重命名项目（Owner）")
async def rename_project(
    project_id: int,
    _owner: OwnerDep,
    payload: Annotated[dict[str, Any], Body(...)],
) -> dict[str, Any]:
    """**只允许修改项目名**（``name``，去空白后 1–200 字符）。

    刻意不开放 status / current_iteration / 关联字段：这些由流水线状态机与决策留痕
    维护，从管理端直改会绕过状态机（非法迁移）并破坏审计链。
    """
    name = str(payload.get("name") or "").strip()
    if not name:
        raise _error(
            status.HTTP_422_UNPROCESSABLE_ENTITY, "validation_error", "name 必填", payload
        )
    if len(name) > 200:
        raise _error(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "validation_error",
            "name 长度不得超过 200",
            {"length": len(name)},
        )

    async with _session_factory()() as session:
        project = (
            await session.execute(select(Project).where(Project.id == project_id))
        ).scalar_one_or_none()
        if project is None:
            raise _error(
                status.HTTP_404_NOT_FOUND, "project_not_found", f"project {project_id} 不存在"
            )
        previous = project.name
        project.name = name
        project.updated_at = func.now()
        await session.commit()
        # 先提交再 refresh：`updated_at` 由服务端生成，flush 后属性会过期，
        # 在异步会话里直接读过期属性会触发隐式 IO（MissingGreenlet → 500）。
        await session.refresh(project)
        result = _project_dict(project)

    logger.info("project_renamed project_id=%s from=%s to=%s", project_id, previous, name)
    return result


__all__ = [
    "MODES",
    "PROJECT_STATUSES",
    "create_project",
    "get_project",
    "list_projects",
    "rename_project",
    "router",
]

File: frontend/src/api/projects.ts
/**
 * Copyright 2026 SciLoop contributors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * 项目接口封装：列表走 store（session.projects），创建走这里。
 * 视图层不直接 fetch（沿用 api/client.ts 的统一请求与错误体解析）。
 */

import { get, patch, post, type ApiError } from './client'

export interface CreateProjectInput {
  name: string
  /** 备注（原「一句话研究问题」，落库在 settings.note） */
  note: string
  fields: string[]
  mode?: 'manual' | 'auto'
}

/** `GET /projects/{id}` 的响应（含迭代历史；缺值一律 null，不编造） */
export interface ProjectDetail {
  id: number
  name: string
  status: string | null
  mode: string | null
  current_iteration: number
  is_demo: boolean
  idea_id: number | null
  taskbook_id: number | null
  created_at: string | null
  updated_at: string | null
  settings?: Record<string, unknown> | null
  stage_order?: string[]
  run_count?: number
  latest_run?: Record<string, unknown> | null
  iteration_history?: Array<Record<string, unknown>>
}

/** 项目详情（公开只读） */
export async function fetchProjectDetail(id: number): Promise<ProjectDetail> {
  return get<ProjectDetail>(`/projects/${id}`)
}

/** 重命名项目：`PATCH /projects/{id}`（Owner 写操作，只改 name） */
export async function renameProject(id: number, name: string): Promise<ProjectDetail> {
  return patch<ProjectDetail>(`/projects/${id}`, { body: { name: name.trim() } })
}

/**
 * 归档 / 取消归档项目：`PATCH /projects/{id}`（Owner 写操作，只改 archived）。
 *
 * 归档 = 从左栏「项目」收起、进「已归档」折叠区；**不动流水线状态、不删数据**，
 * 项目下的对话一并收起，但点开仍可继续聊。
 */
export async function setProjectArchived(id: number, archived: boolean): Promise<ProjectDetail> {
  return patch<ProjectDetail>(`/projects/${id}`, { body: { archived } })
}

export interface CreatedProject {
  id: number
  name: string
  status?: string
  mode?: string
  is_demo?: boolean
}

/**
 * 可选研究领域（内置前 9 个平铺展示；更多方向在弹窗里用下拉选择）。
 * 与后端抓取用的 arXiv 分类同口径（后端只做 JSONB 落库，不校验取值）。
 */
export const RESEARCH_FIELDS: Array<{ value: string; label: string }> = [
  { value: 'cs.AI', label: 'cs.AI · 人工智能' },
  { value: 'cs.CL', label: 'cs.CL · 计算语言学' },
  { value: 'cs.CV', label: 'cs.CV · 计算机视觉' },
  { value: 'cs.LG', label: 'cs.LG · 机器学习' },
  { value: 'cs.IR', label: 'cs.IR · 信息检索' },
  { value: 'cs.SE', label: 'cs.SE · 软件工程' },
  { value: 'cs.DB', label: 'cs.DB · 数据库' },
  { value: 'cs.HC', label: 'cs.HC · 人机交互' },
  { value: 'cs.MA', label: 'cs.MA · 多智能体' },
  { value: 'cs.NE', label: 'cs.NE · 神经与进化计算' },
  { value: 'cs.RO', label: 'cs.RO · 机器人学' },
  { value: 'cs.CR', label: 'cs.CR · 密码学与安全' },
  { value: 'stat.ML', label: 'stat.ML · 统计机器学习' },
  { value: 'eess.AS', label: 'eess.AS · 语音与音频' },
  { value: 'q-bio.BM', label: 'q-bio.BM · 生物分子' },
  { value: 'econ.EM', label: 'econ.EM · 计量经济' },
]

/**
 * 创建项目：`POST /projects`（Owner 写操作）。
 *
 * `settings` 落库：备注 + 研究方向 + 预算阈值（护栏 8.0 / 演示配额 3.0 / 迭代 3 轮，
 * 与 contracts.guardrails.cost 一致；弹窗里不再让用户填，直接带默认值）。
 */
export async function createProject(input: CreateProjectInput): Promise<CreatedProject> {
  return post<CreatedProject>('/projects', {
    body: {
      name: input.name.trim(),
      mode: input.mode ?? 'manual',
      settings: {
        note: input.note.trim(),
        fields: input.fields,
        budget: {
          max_llm_cost_usd: 8.0,
          demo_cost_quota_usd: 3.0,
          max_iterations: 3,
        },
      },
    },
  })
}

/** 判断是否为「只读面被拒」：public_demo 下写操作必然 403 */
export function isOwnerRequired(error: unknown): boolean {
  const status = (error as ApiError | undefined)?.status
  return status === 403
}
